feat(vitest): expose execArgv to the different pools

adriencaccia opened the pull request:
### Description

Fixes #4382

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

AriPerkkio commented:
I like the idea. Maybe add some tests to `test/run/test/`, e.g. `exec-args.test.ts`.

adriencaccia commented:
> I like the idea. Maybe add some tests to `test/run/test/`, e.g. `exec-args.test.ts`.

I added some tests for the base usage with the different pools.

I wanted to add a test when a user is setting an execution arg manually without setting the corresponding config, which would not be passed. But I struggled to create such a test because it would require passing execArgs to the `vitest` command with `runVitestCli`, which is not possible atm.

Do you have any guidance on that?

AriPerkkio commented on test/run/test/exec-args.test.ts:
```diff
@@ -0,0 +1,13 @@
+import { describe, expect, it } from 'vitest'
+import { runVitestCli } from '../../test-utils'
+
+describe('exec-args', async () => {
```
These tests work and look good, but require quite much setup, e.g. 3x `vitest.config.ts`. Instead, you could create test cases to `fixtures/` and use `runVitest` (not cli) from test utils. Something like `test/env-mixed/test/mixed-environments.test.ts` does.

This might work:

```ts
import { expect, test } from 'vitest'
import { runVitest } from '../../test-utils'

test.each(['forks', 'threads', 'vmThreads'] as const)('should pass execArvs to { pool: "%s" } ', async (pool) => {
  const vitest = await runVitest({
    include: ['fixtures/exec-args.test.ts'],
    pool,
    poolOptions: {
      [pool]: {
        execArgv: [
          '--hash-seed=1',
          '--random-seed=1',
          '--no-opt',
        ],
      },
    },
  })

  expect(vitest.stdout).toContain('✓ exec-args-fixtures')
})

```

adriencaccia replied:
Thanks that worked perfectly!

AriPerkkio reviewed (commented):
> But I struggled to create such a test because it would require passing execArgs to the `vitest` command with `runVitestCli`, which is not possible atm.
>
> Do you have any guidance on that?

I think you could use `execa` directly for this. Here's a reference:

https://github.com/vitest-dev/vitest/blob/b3bc866d441eeead32ab5503d2f9565e25c0a0d1/test/browser/specs/runner.test.mjs#L8

adriencaccia commented on test/run/test/exec-args.test.ts:
```diff
@@ -0,0 +1,30 @@
+})
+
+test('should not pass execArgv to workers when not specified in the config', async () => {
+  const { stdout } = await execa('npx', ['vitest', '--run', '--root', 'no-exec-args-fixtures', '--config', 'vitest.config.ts'],
+    { env: { NODE_OPTIONS: '--inspect-brk' } })
```
This does not work as expected, `--inspect-brk` is not passed. I will try and find a way to populate `process.execArgv` when using `npx`.

adriencaccia replied:
As there was no way of doing it here, I modified the loaded `vitest.config.ts` to add the flag that we do not want to be passed

adriencaccia commented:
@AriPerkkio It should be good now!

AriPerkkio commented on test/run/test/exec-args.test.ts:
```diff
@@ -0,0 +1,29 @@
+  expect(vitest.stdout).toContain(`✓ ${fileToTest}`)
+})
+
+test('should not pass execArgv to workers when not specified in the config', async () => {
+  const { stdout } = await execa('npx', ['vitest', '--run', '--root', 'no-exec-args-fixtures', '--config', 'vitest.config.ts'])
+
+  expect(stdout).toContain('✓ no-exec-argv.test.ts')
+})
```
Let's change this to following:

```ts
test('should not pass execArgv to workers when not specified in the config', async () => {
  const { stdout, stderr } = await execa('node', [
    '--title', 'this-works-only-on-main-thread',
    './node_modules/vitest/vitest.mjs', '--run',
    '--root', 'no-exec-args-fixtures',
  ], {
    reject: false,
    env: {
      VITE_NODE_DEPS_MODULE_DIRECTORIES: '/node_modules/,/packages/',
    },
  })

  expect(stderr).not.toContain('Error: Initiated Worker with invalid execArgv flags: --title')
  expect(stderr).not.toContain('ERR_WORKER_INVALID_EXEC_ARGV')
  expect(stdout).toContain('✓ no-exec-argv.test.ts')
})
```

When the implementation breaks, this tests shows following:

```
 FAIL  test/exec-args.test.ts > should not pass execArgv to workers when not specified in the config
AssertionError: expected '\n⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯\nError: Initiated Worker with invalid execArgv flags: --title ... Serialized Error: { code: \'ERR_WORKER_INVALID_EXEC_ARGV\' }\n' to not include 'Error: Initiated Worker with invalid execArgv flags: --title'
 ❯ test/exec-args.test.ts:37:22
     35|   })
     36| 
     37|   expect(stderr).not.toContain('Error: Initiated Worker with invalid execArgv flags: --title')
       |                      ^
     38|   expect(stderr).not.toContain('ERR_WORKER_INVALID_EXEC_ARGV')
     39|   expect(stdout).toContain('✓ no-exec-argv.test.ts')
```

AriPerkkio replied:
The environment variable `VITE_NODE_DEPS_MODULE_DIRECTORIES` is from `.npmrc` which is automatically picked when running scripts through `npm`. But as this test would invoke `node` directly it needs to be added here manually.

https://github.com/vitest-dev/vitest/blob/b3bc866d441eeead32ab5503d2f9565e25c0a0d1/.npmrc#L5

adriencaccia replied:
That works well!
Thank you for the idea and the explanation

AriPerkkio commented on test/run/no-exec-args-fixtures/vitest.config.ts:
This file can be removed now that argv is passed from execa.

adriencaccia replied:
If we do not keep this file, the `test/run/vitest.config.ts` will be picked up as the config, making the test fail because of `include: ['test/**/*.test.*'],`.
Keeping it and having it empty makes sure that `test/run/no-exec-args-fixtures/no-exec-argv.test.ts` will be ran.

AriPerkkio replied:
I see, that's right. Let's keep this file as is.

AriPerkkio commented on packages/vitest/src/types/pool-options.ts:
```diff
@@ -81,6 +81,17 @@ interface WorkerContextOptions {
    */
   isolate?: boolean
+
+  /**
+  * Pass additional arguments to `node` process when spawning `worker_threads` or `child_process`.
+  *
+  * This can be used to enable profile, e.g. `--cpu-prof`. Set to `process.execArgv` to pass all arguments of the current process.
+  *
+  * Be careful when using, it as some options may crash worker, e.g. --prof, --title. See https://github.com/nodejs/node/issues/41103
+  *
+  * @default [] // no execution arguments are passed
+  */
```
```suggestion
  /**
  * Pass additional arguments to `node` process when spawning `worker_threads` or `child_process`.
  *
  * See [Command-line API | Node.js](https://nodejs.org/docs/latest/api/cli.html) for more information.
  *
  * Set to `process.execArgv` to pass all arguments of the current process.
  *
  * Be careful when using, it as some options may crash worker, e.g. --prof, --title. See https://github.com/nodejs/node/issues/41103
  *
  * @default [] // no execution arguments are passed
  */
```

AriPerkkio commented on docs/config/index.md:
```diff
@@ -836,6 +858,17 @@ Use Atomics to synchronize threads.
+- **Type:** `string[]`
+- **Default:** `[]`
+
+Pass additional arguments to `node` process in the VM context.
```
```suggestion
Pass additional arguments to `node` process in the VM context. See [Command-line API | Node.js](https://nodejs.org/docs/latest/api/cli.html) for more information.

```

AriPerkkio commented on test/run/exec-args-fixtures/forks.test.ts:
```diff
@@ -0,0 +1,14 @@
+
+describe('exec-args', async () => {
+  it('should have the correct flags', () => {
+    // passed via poolOptions.forks.execArgv in vitest.config.ts
```
There's no `vitest.config.ts` anymore, so:

```suggestion
```

AriPerkkio commented on test/run/exec-args-fixtures/threads.test.ts:
```diff
@@ -0,0 +1,12 @@
+
+describe('exec-args', async () => {
+  it('should have the correct flags', () => {
+    // passed via poolOptions.forks.execArgv in vitest.config.ts
```
```suggestion
```

AriPerkkio commented on test/run/exec-args-fixtures/vmThreads.test.ts:
```diff
@@ -0,0 +1,16 @@
+
+describe('exec-args', async () => {
+  it('should have the correct flags', () => {
+    // passed via poolOptions.vmThreads.execArgv in vitest.config.ts
```
```suggestion
```

AriPerkkio commented on test/run/no-exec-args-fixtures/no-exec-argv.test.ts:
```diff
@@ -0,0 +1,12 @@
+describe('exec-args', async () => {
+  it('should have the correct flags', () => {
+    // flags should not be passed
+    expect(process.execArgv).not.toContain('--inspect-brk')
```
```suggestion
    expect(process.execArgv).not.toContain('--title')
```

AriPerkkio reviewed (commented):
Last small changes and I think this is good to go. 

I think it would also be good to have an example of how to profile tests using these new options in https://vitest.dev/guide/debugging.html#debugging. But that can be left out of this PR if it's too much for now.

AriPerkkio commented on test/run/test/exec-args.test.ts:
```diff
@@ -0,0 +1,40 @@
+    '--root', 'no-exec-args-fixtures', '--dir', 'no-exec-args-fixtures',
+  ], {
+    reject: false,
+    env: {
+      VITE_NODE_DEPS_MODULE_DIRECTORIES: '/node_modules/,/packages/',
+    },
```
```suggestion
    env: {
      VITE_NODE_DEPS_MODULE_DIRECTORIES: '/node_modules/,/packages/',
      NO_COLOR: '1',
    },
```

AriPerkkio reviewed (approved):
Looks good. I did some manual testing on reference project and it works well:

```ts
import { defineConfig } from "vitest/config";

export default defineConfig({
  test: {
    pool: "threads",
    poolOptions: {
      threads: {
        execArgv: ["--cpu-prof", "--cpu-prof-dir=./profiling"],
      },
    },
  },
});
```

Thanks!

sheremet-va commented:
Maybe we should still filter `execArgv` if `poolOptions` are not provided?

AriPerkkio commented:
With the current changes the Node CLI arguments are not passed to workers, ever. So no filtering is needed. 

Earlier we only passed `--cpu-prof` and `--heap-prof` CLI options to workers. @sheremet-va do you mean that we should keep continuing to pass those from main thread to workers as before?

sheremet-va commented:
> @sheremet-va do you mean that we should keep continuing to pass those from main thread to workers as before? Otherwise there is no need for filtering the options at all.

Yes, this is what I mean.

adriencaccia commented:
@sheremet-va, @AriPerkkio I added back the original filtering, and adding `--diagnostic-dir` go through as well since it is strongly tied to the already allowed ones.
I added a test to make sure that the flags are passing through.

Let me know if I need to change something in the PR description/commit messages, or If you want me to rebase the third commit into the first one, to not have the deletion of the filtering and then adding it back.

AriPerkkio commented:
Breaking changes of PR description can now be removed.

adriencaccia commented:
> Breaking changes of PR description can now be removed.

Done ✅ 

There is also a `BREAKING CHANGE:` footer in the first commit. Should I remove that too?

AriPerkkio commented:
> I do not know what will be picked up for the changelog/release, the PR description or the commits

Commit title and message body can be changed when maintainers merge this PR so we'll just need to remember to remove it then.
